iothub/device/src/InternalClient.cs

```diff
@@ -879,7 +879,18 @@ public async Task SetMethodHandlerAsync(string methodName, MethodCallback method
             }
             finally
             {
-                _methodsDictionarySemaphore.Release();
+                try
+                {
+                    _methodsDictionarySemaphore.Release();
+                }
+                catch (SemaphoreFullException e)
+                {
+                    // this semaphore is typically grabbed at the start of the method, but if
+                    // this method is cancelled while waiting to grab the semaphore, then this semaphore.release
```

Comment:
Good catch!

A question - 
"this method is cancelled while waiting to grab the semaphore" -> does this mean that the cancellation token passed expires before grabbing the semaphore?
would that not throw an OperationCancelledException, or something along those lines?

Also, wouldn't it be better to do a cancellationToken.ThrowIfCancellationRequested instead?

Reply:
> does this mean that the cancellation token passed expires before grabbing the semaphore?

Yes, or the cancellation token expires during the wait for the semaphore

> would that not throw an OperationCancelledException...?

Regardless of what it throws, these finally blocks will execute. So the idea is that the finally blocks should try to release the semaphore, but shouldn't throw if they didn't have the semaphore to begin with

Reply:
> wouldn't it be better to do a cancellationToken.ThrowIfCancellationRequested instead?

I don't follow. We give the cancellation token to the semaphore during ```semaphore.WaitAsync(cancellationToken)``` so what control do we have over that?

Reply:
The added catch statement for SemaphoreFullException basically just allows for the ```semaphore.release()``` call to "always work." If the semaphore was grabbed, then the release is successful and the catch block isn't hit. If the semaphore wasn't grabbed, then the release will throw this SemaphoreFullException and this catch block will prevent it from reaching the user, but importantly the release wasn't necessary anyways since the semaphore wasn't grabbed

Reply:
so there are 3 possible broad scenarios:
- cancellation token was valid throughout the duration of the method call -> the semaphore was acquired and then released as expceted
- cancellation token expired during method processing -> semaphore acquired, threw OperationCancelledException when the cancellation token expired, semaphore released.
- cancellation token was already expired -> semaphore not acquired, so the release had nothing to release -> threw exception.

My question is around scenario 3 -> if the cancellation token was already expired, wouldn't it make more sense to call ThrowIfCancellationRequested before we use that cancellation token (be it either in passing it into the semaphore, or into any other subsequent method calls).

Reply:
> wouldn't it make more sense to call ThrowIfCancellationRequested before we use that cancellation token

Just as an addition to these changes? Sure that sounds reasonable. I'll add it